pkg/ccl/backupccl/backup_test.go

```diff
@@ -11204,3 +11204,44 @@ func TestRestoreMemoryMonitoringMinWorkerMemory(t *testing.T) {
 	actualFingerprints := sqlDB.QueryStr(t, "SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE restore.bank")
 	require.Equal(t, expectedFingerprints, actualFingerprints)
 }
+
+// TestBoundBackupExportOnRanges tests that export requests never get sent
+// across range boundaries, even if the ranges are empty.
+func TestBoundBackupExportOnRanges(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	const numAccounts = 1
+
+	waitForSplits := make(chan struct{})
+	var responseCount atomic.Int32
+	params := base.TestClusterArgs{}
+	knobs := base.TestingKnobs{
+		DistSQL: &execinfra.TestingKnobs{
+			BackupRestoreTestingKnobs: &sql.BackupRestoreTestingKnobs{
+				RunBeforeBackupFlow: func(ctx context.Context) {
+					<-waitForSplits
+				},
+				RunAfterExportingSpanEntry: func(ctx context.Context, response *kvpb.ExportResponse) {
+					responseCount.Add(1)
+				},
```

Comment:
Just in case, should we be actually adding the length of the `files` portion of the response to make sure we don't see two files in one response.

Reply:
I'm not sure if that always asserts the property we want, namely that we only called evalExport once to get the response? I'd have to go look at empty responses and o.combine again, but I thought empty responses sent no file, not an empty file.

This diff as it stands is testing that we got a certain number of replies from a distsender call that sent an export requests. 

But the issue that the we had before which was uncaught by out tests was that a single request sent would be served by many ranges, whereas we thought it would be served by one, and we aren't actually testing _that_ attribute here, in this test.

If our goal is to just test how many requests we intend to send (potentially by counting how many we get replies for as we do here), then this is OK, though we don't actually need to send the requests at all: _that_ behavior we could _unit_ test: it is essentially `splitSpans` but we could pull the entire thing into a testable function that given `BackupProcessorSpec` produces the `todo` chan, and then unit test that function.

I think this test is sorta indirectly doing that, by counting req span loop iterations, and comparing that to the range count. If that's our goal -- just that we send the expected number of reqs -- then I'm not sure we need to look at Files.

I think I can maybe convince myself that a test that checks that we have at least as many req span loop iterations as we had ranges is a step in the right direction.

But my original desire, when I said we needed a new test in this area is to assert the actual number of times we call `evalExport` per request we make:  _that_ is what the underlying issue was, and was uncaught in our previous tests. And I don't think we're actually really checking _that_ in this test either; We're somewhat indirectly asserting we'll make at least as a number of requests as there are ranges, which I think _implies_ we won't make multiple evalExport calls per request, but we're not actually testing how many we make.

The approach I had in mind when I said I wanted to write this test was to compare the trace recording of each export response to count how many times we see the req limiter and/or evalExport span, so that we could assert our splitting is getting the actual underlying kv behavior we think our splitting gets us, namely one call to evalExport per request sent. I think if since we pass ctx to RunAfterExportingSpanEntry we can grab the trace recording there?


I think we also want this test to have a mixture of empty and non-empty ranges, and a mixture of incremental and non-incremental requests, including some incremental requests over adjacent ranges that have no changes in the target time period, since _that_ is the case that has really bitten us.

_All of that said_, directionally this test as it stands is an improvement,  in that it is getting us more coverage of `splitSpans` than we previously had, so I don't have any qualms about merging it now as it is but I wouldn't consider it as completing 115575 without verifying actual evalExport calls made. I do think though that what it is testing now -- correctly producing the right number of request spans -- it could do so by _unit_ testing a function from BackupDataSpec to `todo` chan.

Reply:
> If our goal is to just test how many requests we intend to send (potentially by counting how many we get replies for, as we do here), then this is OK, though we don't actually need to send the requests at all: that behavior we could unit test: it is essentially splitSpans but we could pull the entire thing into a testable function that given BackupProcessorSpec produces the todo chan, and then unit test that function.

I considered doing that, but since I want to backport this test, I didn't want to change prod code for a test.

> we could assert our splitting is getting the actual underlying kv behavior we think our splitting gets us, namely one request is served by one one call to evalExport.

Good point. Lemme see if I can grab this while minimizing prod code changes.

Reply:
> I think we also want this test to have a mixture of empty and non-empty ranges, and a mixture of incremental and non-incremental requests, including some incremental requests over adjacent ranges that have no changes in the target time period, since that is the case that has really bitten us.

I like that idea. So the high level invariant is: one export request per range maps to one `evalExport` call, regardless if we're running full or inc backups, and if there's a bit of data in the range. For the scope of this test, I'm not going create a workload that would induce export request to return early within a range.

Reply:
I've opened a new PR to unit test the backup processor logic here https://github.com/cockroachdb/cockroach/pull/115902. I'll continue to work on this PR to write a test around `EvalExport` calls.